File: js/ev2/ut5-poo-colecciones/03-objetos-y-clases/practicas/madrid-cesjuanpablosegundo-figuras/figuras-resuelto.js
class Figura { calcularArea(){ throw new Error("abstract"); } mostrarDatos(){ console.log(`Area = ${this.calcularArea()}`); } }
class Circulo extends Figura { constructor(radio){ super(); this.radio=radio; } calcularDiametro(){ return 2*this.radio; } calcularArea(){ return Math.PI*this.radio*this.radio; }
  mostrarDatos(){ console.log("Tipo = Circulo"); console.log(`Radio = ${this.radio}`); console.log(`Diametro = ${this.calcularDiametro()}`); super.mostrarDatos(); } }
class Rectangulo extends Figura { constructor(base,altura){ super(); this.base=base; this.altura=altura; } calcularPerimetro(){ return 2*(this.base+this.altura); } calcularArea(){ return this.base*this.altura; }
  mostrarDatos(){ console.log("Tipo = Rectangulo"); console.log(`Base = ${this.base}`); console.log(`Altura = ${this.altura}`); console.log(`Perimetro = ${this.calcularPerimetro()}`); super.mostrarDatos(); } }
class Triangulo extends Figura { constructor(base,altura){ super(); this.base=base; this.altura=altura; } calcularArea(){ return (this.base*this.altura)/2; }
  mostrarDatos(){ console.log("Tipo = Triangulo"); console.log(`Base = ${this.base}`); console.log(`Altura = ${this.altura}`); super.mostrarDatos(); } }
const figuras = [new Circulo(3), new Rectangulo(2,4), new Triangulo(3,6)];
figuras.forEach(f => f.mostrarDatos());
console.log(`Area total = ${figuras.reduce((a,f)=>a+f.calcularArea(),0)}`);

File: cpp/src/ev2/ut6_pooavanzadaestructuras/u01herenciapolimorfismo/practicas/madridcesjuanpablosegundofiguras/figuras_resuelto.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <cmath>
using namespace std;
class Figura { public: virtual ~Figura()=default; virtual double calcularArea()=0;
  virtual void mostrarDatos(){ cout << "Area = " << calcularArea() << "\n"; } };
class Circulo: public Figura { public: double radio; Circulo(double r):radio(r){}
  double diametro(){ return 2*radio; } double calcularArea() override { return M_PI*radio*radio; }
  void mostrarDatos() override { cout << "Tipo = Circulo\nRadio = "<<radio<<"\nDiametro = "<<diametro()<<"\n"; Figura::mostrarDatos(); } };
class Rectangulo: public Figura { public: double base,altura; Rectangulo(double b,double a):base(b),altura(a){}
  double perimetro(){ return 2*(base+altura); } double calcularArea() override { return base*altura; }
  void mostrarDatos() override { cout << "Tipo = Rectangulo\n"; Figura::mostrarDatos(); } };
class Triangulo: public Figura { public: double base,altura; Triangulo(double b,double a):base(b),altura(a){}
  double calcularArea() override { return (base*altura)/2; }
  void mostrarDatos() override { cout << "Tipo = Triangulo\n"; Figura::mostrarDatos(); } };
int main(){ vector<unique_ptr<Figura>> f; f.push_back(make_unique<Circulo>(3)); f.push_back(make_unique<Rectangulo>(2,4)); f.push_back(make_unique<Triangulo>(3,6));
  double t=0; for(auto& x:f){ x->mostrarDatos(); t+=x->calcularArea(); } cout<<"Area total = "<<t<<"\n"; return 0; }
